tagstudio/src/core/library.py

```diff
@@ -2324,3 +2172,287 @@ def sort_fields(self, entry_id: int, order: list[int]) -> None:
+
+        # For each verified, extracted Tag term.
+        failure_to_union_terms = False
+        for term in all_tag_terms:
+            # If the term from the previous loop was already verified:
+            if not failure_to_union_terms:
+                cluster: set = set()
+                # Add the immediate associated Tags to the set (ex. Name, Alias hits)
+                # Since this term could technically map to multiple IDs, iterate over it
+                # (You're 99.9999999% likely to just get 1 item)
+                for id in self._tag_strings_to_id_map[term]:
+                    cluster.add(id)
+                    cluster = cluster.union(set(self.get_tag_cluster(id)))
+                # For each of the Tag IDs in the term's ID cluster:
+                for t in cluster:
+                    # Assume that this ID from the cluster is not in the Entry.
+                    # Wait to see if proven wrong.
+                    failure_to_union_terms = True
+                    # If the ID actually is in the Entry,
+                    if t in entry_tags:
+                        # There wasn't a failure to find one of the term's cluster IDs in the Entry.
+                        # There is also no more need to keep checking the rest of the terms in the cluster.
+                        failure_to_union_terms = False
+                        # print(f"FOUND MATCH: {t}")
+                        break
+                    # print(f'\tFailure to Match: {t}')
+        # # failure_to_union_terms is used to determine if all terms in the query were found in the entry.
+        # # If there even were tag terms to search through AND they all match an entry
+        if all_tag_terms and not failure_to_union_terms:
+            return add_entry(entry)
+        else:
+            return None
+
+    def add_entries_from_special_flags(
+        self,
+        special_flags: tuple[bool, bool, bool, bool],
+        entry_tags: list[int],
+        entry_authors: list[list[str]],
+        entry: Entry,
+    ):
+        (only_untagged, only_no_author, only_empty, only_missing) = special_flags
+        if only_untagged and not entry_tags:
+            return True
+        elif only_no_author and not entry_authors:
+            return True
+        elif only_empty and (
+            len(entry.fields) <= 0 or len(entry.fields[0].keys()) <= 0
+        ):
+            return True
+        elif (
+            only_missing
+            and (self.library_dir / entry.path / entry.filename).resolve()
+            in self.missing_files
+        ):
+            return True
+        return False
+
+    def remap_fields(self, entry: Entry) -> dict[str | int, str]:
+        entry_fields: dict = {}
+        for field in entry.fields:
+            if len(field.keys()) > 0:
+                field_keys = list(field.keys())
+                field_key = self._field_id_to_name_map.get(int(field_keys[0]))
+                entry_fields[field_key] = list(field.values())[0]
```

Comment:
This is breaking change. Currently each item can have the same field multiple times. I'm not sure if the design is intentional or accidental, @CyanVoxel? 

With this change you take only the first occurance.


identical code for this could be a bit simplifed
```suggestion
            for key, val in field.items():
                key_mapped = self._field_id_to_name_map.get(int(key))
                entry_fields[key_mapped] = val
```

Reply:
Right, it's intentional design that there can be more than one field of a single type applied to an entry. For example, two author fields with two different authors.